include/fiction/algorithms/simulation/sidb/operational_domain.hpp

```diff
@@ -377,8 +379,27 @@ class operational_domain_impl
             output_bdl_pairs{detect_bdl_pairs<Lyt>(
                 layout, sidb_technology::cell_type::OUTPUT,
                 ps.operational_params.input_bdl_iterator_params.bdl_wire_params.bdl_pairs_params)},
-            num_dimensions{params.sweep_dimensions.size()}
+            num_dimensions{params.sweep_dimensions.size()},
+            input_bdl_wires{detect_bdl_wires(lyt, params.operational_params.input_bdl_iterator_params.bdl_wire_params,
+                                             bdl_wire_selection::INPUT)},
+            output_bdl_wires{detect_bdl_wires(lyt, params.operational_params.input_bdl_iterator_params.bdl_wire_params,
+                                              bdl_wire_selection::OUTPUT)}
     {
+        if (lyt.has_cell_type(technology<Lyt>::cell_type::LOGIC))
+        {
+            lyt.foreach_cell(
+                [&](const auto& c)
+                {
+                    if (lyt.get_cell_type(c) == technology<Lyt>::cell_type::LOGIC)
+                    {
+                        canvas_lyt.assign_cell_type(c, technology<Lyt>::cell_type::NORMAL);
+                    }
+                });
+        }
```

Comment:
There seems to be a design flaw here. You extra created the function `has_cell_type` to check for logic cells, which iterates over the entire layout in the worst case. However, you discard the information of the cells and then iterate again to gather them. You might want to rethink the function `has_cell_type` in favor of a `get_cells_by_type` (or something) function that returns all cells of a certain type. That eliminates the need for a second iteration.